### ISSUE TITLE
About using WT convolution directly on the original RGB image

### ISSUE THREAD

**Issue by HDL-YD** [BGU-CS-VIL/WTConv#41]
Dear sir，I am trying to use WT convolution on the original image. Accumulate the results obtained from the operation onto the original image. But after visualizing the results, I found that the colors in the entire image were very strange. May I ask if WT convolution cannot be directly used on RGB images.

**Comment by shahaffind** [comment]
Hey @HDL-YD ,
My guess is there might be some numerical issues. If you normalize the image to be within [0,1], try to clip the output to those values as well:
```python
output[output < 0] = 0
output[output > 1] = 1
```
The image printing might rescale it to this range and result in a color shift.

If that doesn't help, please share a code snippet and example images and I will try to debug it

Shahaf

**Comment by HDL-YD** [comment]
Dear sir，the code and image results are shown below.
`import os
from PIL import Image
import torch
import torchvision.transforms as transforms

from wtconv import WTConv2d

def set_seed(seed=42):
    torch.manual_seed(seed)
    if torch.cuda.is_available():
        torch.cuda.manual_seed_all(seed)

def apply_convolution_rgb(input_folder, output_folder, wt_levels=3, kernel_size=5):
    """
    对输入文件夹中的所有RGB图片直接应用卷积操作，并将结果与原图叠加，保存为新的RGB图片。
    
    :param input_folder: 输入图片文件夹路径
    :param output_folder: 输出图片保存文件夹路径
    :param wt_levels: 小波变换层数，默认为3
    :param kernel_size: 卷积核大小，默认为5
    """
    set_seed()
    
    if not os.path.exists(output_folder):
        os.makedirs(output_folder)
    
    img_extensions = ['.jpg', '.jpeg', '.png', '.bmp', '.tiff']
    
    preprocess = transforms.ToTensor()
    to_pil = transforms.ToPILImage()

    with torch.no_grad(): 
        for filename in os.listdir(input_folder):
            if any(filename.lower().endswith(ext) for ext in img_extensions):
                filepath = os.path.join(input_folder, filename)
                
                img = Image.open(filepath).convert('RGB')
                
                img_tensor = preprocess(img)
                
                in_chs = out_chs = 3
                conv_layer = WTConv2d(in_chs, out_chs, kernel_size=kernel_size, wt_levels=wt_levels)
                
                conv_result = conv_layer(img_tensor.unsqueeze(0)).squeeze(0)   # 不添加batch维度
                
                conv_result = torch.clamp(conv_result, min=0, max=1)
                
                combined_result = 0.5*img_tensor + 0.5*conv_result
                
                combined_result = torch.clamp(combined_result, min=0, max=1)
                
                combined_result = (combined_result * 255).to(torch.uint8)
                
                final_img = to_pil(combined_result)
                
                final_img.save(os.path.join(output_folder, filename))

input_dir = '/home/hdl/UODReview/DUO/test_origin' 
output_dir = '/home/hdl/UODReview/DUO/test_WT'  
apply_convolution_rgb(input_dir, output_dir)
`
img1 

![Image](https://github.com/user-attachments/assets/b9507d4c-6be9-4cb6-94e5-00696c5ae5b9)

img1-result

![Image](https://github.com/user-attachments/assets/0a673b47-2732-4b2e-b89e-7e838456b677)

img2

![Image](https://github.com/user-attachments/assets/f5f440e5-064f-4e0b-a950-071c0c42e269)

img2-result

![Image](https://github.com/user-attachments/assets/ec6a8466-eeab-4fe6-8245-4e99900709d1)

When comparing the resultant image to the original, the former appears to exhibit only an additional layer of color overlay, while all other image characteristics (e.g., structural details, textures, and compositional elements) remain entirely unchanged.‌
I am confused and would appreciate your assistance in resolving this issue. Thank you very much for your support. ‌

**Comment by shahaffind** [comment]
Oh, I thought you were only talking about the WT operation.
It is normal that the output of a convolution will look different from the input, and since the input is a 4k-resolution image, the pixel-wise shuffling will not be too visible. Consider kernel_size=5 and wt_levels=3, you have a 40x40 pixels of receptive field in a 3840x2160 pixels area.
To see a difference, you either need to stack a few layers together, or increase (either or both) the kernel_size and wt_levels.
You can try a normal convolution and see similar results (even when the kernel size is set pretty high)
